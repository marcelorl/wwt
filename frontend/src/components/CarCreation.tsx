import { Box, Button, Modal, Typography } from '@mui/material';
import { useState } from 'react';
import { Form, Formik } from 'formik';
import * as Yup from 'yup';
import { FieldFormGroup } from './FieldFormGroup';
import { Car } from '../types/general';
import { useCreateCar } from '../hooks/useCreateCar';

const CarSchema = Yup.object().shape({
  make: Yup.string().required('Required'),
  model: Yup.string().required('Required'),
  color: Yup.string().required('Required'),
  year: Yup.number().max(new Date().getFullYear(), 'Invalid year'),
  category: Yup.string().required('Required'),
  mileage: Yup.number().required('Required'),
  price: Yup.number().required('Required'),
});

const style = {
  position: 'absolute',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)',
  width: 400,
  bgcolor: 'background.paper',
  boxShadow: 24,
  p: 4,
};

type Props = {
  isOpen: boolean;
  setIsOpen: (o: boolean) => void;
};

export const CarCreation = ({ isOpen, setIsOpen }: Props) => {
  const mutation = useCreateCar(() => setIsOpen(false));

  return (
    <>
      <Button onClick={() => setIsOpen(true)}>Add Car</Button>
      <Modal
        open={isOpen}
        onClose={() => setIsOpen(false)}
        aria-labelledby="modal-modal-title"
        aria-describedby="modal-modal-description"
      >
        <Box sx={style}>
          <Formik
            initialValues={{
              make: '',
              model: '',
              color: '',
              year: 0,
              category: '',
              mileage: 0,
              price: 0,
            }}
            validationSchema={CarSchema}
            onSubmit={(values) => {
              mutation.mutate(values as Car);
            }}
          >
            {({ errors, touched }) => (
              <Form>
                <Typography id="modal-modal-title" variant="h6" component="h2">
                  Create new Car
                </Typography>
                <Box>
                  <FieldFormGroup field="make" />
                  <FieldFormGroup field="model" />
                  <FieldFormGroup field="package" />
                  <FieldFormGroup field="color" />
                  <FieldFormGroup field="year" type="number" />
                  <FieldFormGroup field="category" />
                  <FieldFormGroup field="mileage" type="number" />
                  <FieldFormGroup field="price" type="number" />
                </Box>

                <Button type="submit">Submit</Button>
              </Form>
            )}
          </Formik>
        </Box>
      </Modal>
    </>
  );
};
