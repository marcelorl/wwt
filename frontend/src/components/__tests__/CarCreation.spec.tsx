import React from 'react';
import { describe, test, expect, vi } from 'vitest';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { QueryClient, QueryClientProvider } from 'react-query';
import { CarCreation } from '../CarCreation';

const queryClient = new QueryClient();

const renderComponent = () =>
  render(
    <QueryClientProvider client={queryClient}>
      <CarCreation isOpen setIsOpen={() => vi.fn()} />
    </QueryClientProvider>
  );

describe('#CarCreation', () => {
  test('renders Add Car button and modal', () => {
    const { baseElement, debug } = renderComponent();
    debug();
    expect(baseElement).toMatchSnapshot();
  });

  test('validates form submission', async () => {
    const { getByLabelText, getByText, baseElement } = renderComponent();

    await waitFor(() => userEvent.type(getByLabelText(/year/i), '2030'));
    await waitFor(() => userEvent.click(getByText(/submit/i)));

    expect(baseElement).toMatchSnapshot();
  });
});

//
// test('closes modal on button click', async () => {
//   renderComponent();
//   userEvent.click(screen.getByText(/Add Car/i));
//
//   await waitFor(() => {
//     userEvent.click(screen.getByText(/Submit/i));
//   });
//
//   expect(screen.getByText(/Create new Car/i)).not.toBeInTheDocument();
// });
//
// test('submits form on button click', async () => {
//   renderComponent();
//   userEvent.click(screen.getByText(/Add Car/i));
//
//   const mutationMock = useCreateCar as jest.Mock;
//   const mutateMock = mutationMock.mock.calls[0][0].mutate;
//
//   await waitFor(() => {
//     userEvent.click(screen.getByText(/Submit/i));
//   });
//
//   // You may want to add assertions based on your useCreateCar logic
//   expect(mutateMock).toHaveBeenCalled();
// });
