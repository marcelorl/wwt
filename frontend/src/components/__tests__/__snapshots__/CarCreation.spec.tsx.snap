// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#CarCreation > renders Add Car button and modal 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Add Car
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    aria-describedby="modal-modal-description"
    aria-labelledby="modal-modal-title"
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiBox-root css-1acyv19"
      tabindex="-1"
    >
      <form
        action="#"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          id="modal-modal-title"
        >
          Create new Car
        </h2>
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="make"
            >
              make
            </label>
            <input
              id="make"
              name="make"
              type="text"
              value=""
            />
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="model"
            >
              model
            </label>
            <input
              id="model"
              name="model"
              type="text"
              value=""
            />
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="package"
            >
              package
            </label>
            <input
              id="package"
              name="package"
              type="text"
              value=""
            />
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="color"
            >
              color
            </label>
            <input
              id="color"
              name="color"
              type="text"
              value=""
            />
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="year"
            >
              year
            </label>
            <input
              id="year"
              name="year"
              type="number"
              value="0"
            />
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="category"
            >
              category
            </label>
            <input
              id="category"
              name="category"
              type="text"
              value=""
            />
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="mileage"
            >
              mileage
            </label>
            <input
              id="mileage"
              name="mileage"
              type="number"
              value="0"
            />
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="price"
            >
              price
            </label>
            <input
              id="price"
              name="price"
              type="number"
              value="0"
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Submit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`#CarCreation > validates form submission 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Add Car
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    aria-describedby="modal-modal-description"
    aria-labelledby="modal-modal-title"
    class="MuiModal-root css-79ws1d-MuiModal-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiBox-root css-1acyv19"
      tabindex="-1"
    >
      <form
        action="#"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          id="modal-modal-title"
        >
          Create new Car
        </h2>
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="make"
            >
              make
            </label>
            <input
              id="make"
              name="make"
              type="text"
              value=""
            />
            <div
              style="color: red;"
            >
              Required
            </div>
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="model"
            >
              model
            </label>
            <input
              id="model"
              name="model"
              type="text"
              value=""
            />
            <div
              style="color: red;"
            >
              Required
            </div>
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="package"
            >
              package
            </label>
            <input
              id="package"
              name="package"
              type="text"
              value=""
            />
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="color"
            >
              color
            </label>
            <input
              id="color"
              name="color"
              type="text"
              value=""
            />
            <div
              style="color: red;"
            >
              Required
            </div>
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="year"
            >
              year
            </label>
            <input
              id="year"
              name="year"
              type="number"
              value="2030"
            />
            <div
              style="color: red;"
            >
              Invalid year
            </div>
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="category"
            >
              category
            </label>
            <input
              id="category"
              name="category"
              type="text"
              value=""
            />
            <div
              style="color: red;"
            >
              Required
            </div>
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="mileage"
            >
              mileage
            </label>
            <input
              id="mileage"
              name="mileage"
              type="number"
              value="0"
            />
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              for="price"
            >
              price
            </label>
            <input
              id="price"
              name="price"
              type="number"
              value="0"
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Submit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          >
            <span
              class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
              style="width: 2.8284271247461903px; height: 2.8284271247461903px; top: -1.4142135623730951px; left: -1.4142135623730951px;"
            >
              <span
                class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
              />
            </span>
          </span>
        </button>
      </form>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
